[LLD][ELF][AArch64] Discard .ARM.attributes sections

smithp35 opened the pull request:
LLVM has started to emit AArch64 build attributes sections called .ARM.attributes. LLD does not yet have support for these so they are accumulating in the ELF output. As the first part of that support discard all the .ARM.attributes sections. This can be built upon by the full implementation in LLD.

The build attributes specification only defines build attributes for relocatable objects. The intention for LLD is that files of type ET_EXEC and ET_SHARED will not have a build attributes in the output. A relocatable link with -r will need a merged build attributes, but until the merge is implemented it is better to discard.

smithp35 committed:
[LLD][ELF][AArch64] Discard .ARM.attributes sections

LLVM has started to emit AArch64 build attributes sections called
.ARM.attributes. LLD does not yet have support for these so they are
accumulating in the ELF output. As the first part of that support
discard all the .ARM.attributes sections. This can be built upon by
the full implementation in LLD.

The build attributes specification only defines build attributes for
relocatable objects. The intention for LLD is that files of type
ET_EXEC and ET_SHARED will not have a build attributes in the output.
A relocatable link with -r will need a merged build attributes, but
until the merge is implemented it is better to discard.

sivan-shani commented on lld/test/ELF/aarch64-build-attributes-output.s:
```diff
@@ -0,0 +1,26 @@
+// REQUIRES: aarch64
```
Since the test does not actually requires aarch64 hardware, and also, AARch64 build attributes are relevant also for other platforms (e.g. cross compiling) wouldn't it be better to remove this directive and have the test run an any platform?
There is no certainty that an AArch64 platform will be available for testing.

zmodem replied:
The llc invocation requires aarch64 target support, which I believe is what this directive checks for.

nico replied:
This means "requires an aarch64 backend in LLVM". It's not about the machine running the test.

sivan-shani commented on lld/test/ELF/aarch64-build-attributes-output.s:
```diff
@@ -0,0 +1,26 @@
+/// File has a Build attributes section. This should not appear in
+/// ET_EXEC or ET_SHARED files as there is no requirement for it to
+/// do so. FIXME, the ld -r (relocatable link) should output a single
+/// merged build attributes section. WHen full support is added in
```
typo 'When'

nico reviewed (approved):
Thanks!

zmodem reviewed (approved):
My preference would be to revert the breaking change and find the proper solution (fix all the linkers, apply shf_exclude, put behind a flag..) without time pressure. Especially since it landed before the branch point.

But if you think this is better, please land it so at least lld users get fixed :)

smithp35 commented:
> My preference would be to revert the breaking change and find the proper solution (fix all the linkers, apply shf_exclude, put behind a flag..) without time pressure. Especially since it landed before the branch point.
> 
> But if you think this is better, please land it so at least lld users get fixed :)

FWIW I've just checked GNU ld. It does already have some code for SHT_AARCH64_ATTRIBUTES as the section name has been reserved since the initial AArch64 ABI. This has the same effect as the lld patch above.
https://github.com/bminor/binutils-gdb/blob/master/bfd/elfnn-aarch64.c#L8383 

So I think that leaves us with people building clang alone, and then using an older lld. I'm hopeful that I can add SHF_EXCLUDE to the spec so that we can make sure old LLD does not accumulate the sections.

smithp35 committed:
!fixup fix typo

WHen should be When.

smithp35 commented:
I'll wait to the end of the UK day to see if @MaskRay has any input. Otherwise I'll land it and fix up any suggestions post-commit (with apologies).

MaskRay commented:
> LLVM has started to emit AArch64 build attributes sections called .ARM.attributes. 

Add a link #123990 ?

MaskRay commented on lld/test/ELF/aarch64-build-attributes-output.s:
```diff
@@ -0,0 +1,26 @@
+// REQUIRES: aarch64
```
Perhaps omit `-output` from the name.

smithp35 replied:
ACK.

smithp35 committed:
!fixup rename test to remove output

Rename test/ELF/aarch64-build-attributes-output.s as
test/ELF/aarch64-build-attributes.s

smithp35 commented:
Thanks, was waiting for the buildkite to come back, but has been waiting for Linux for over an hour

zmodem commented:
/cherry-pick ba476d0b83dc8a4bbf066dc02a0f73ded27114f0
